[pytorch] aten codegen to filter backward ops for inference-only mobile build

ljk53 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34276 [pytorch] aten codegen to filter backward ops for inference-only mobile build**
* #34641 [pytorch] aten codegen to filter backends for default mobile build
* #34622 [pytorch] force c10 schema registration for custom build
* #34275 [pytorch] register c10 ops for static dispatch to unblock c10 boxing

Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure mobile build size:
```
// Default mobile build:
scripts/build_pytorch_android.sh armeabi-v7a

// MobileNetV2 custom build:
SELECTED_OP_LIST=MobileNetV2.yaml scripts/build_pytorch_android.sh armeabi-v7a
```

- arm-v7 Android AAR (compressed) size:
```
+----------+-------------------+---------------+
|          | MobileNetV2 Build | Default Build |
+----------+-------------------+---------------+
| Original |         3,354,589 |     5,731,992 |
| #34275   |         3,404,978 |     6,640,526 |
| #34622   |         3,432,569 |     6,640,526 |
| #34641   |         3,431,660 |     6,534,135 |
| This PR  |         3,429,480 |     6,049,387 |
+----------+-------------------+---------------+
```

Differential Revision: [D20266241](https://our.internmc.facebook.com/intern/diff/D20266241)

ljk53 committed:
[pytorch] aten codegen to filter backward ops / backends for inference-only mobile build

Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure Android AAR build size:
```
scripts/build_pytorch_android.sh armeabi-v7a
```

This PR improved: 6.4M -> 5.8M
Baseline (before enabling c10 registers for static dispatch): 5.5M

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] aten codegen to filter backward ops / backends for inference-only mobile build"


Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure Android AAR build size:
```
scripts/build_pytorch_android.sh armeabi-v7a
```

This PR improved: 6.4M -> 5.8M
Baseline (before enabling c10 registers for static dispatch): 5.5M

Differential Revision: [D20266241](https://our.internmc.facebook.com/intern/diff/D20266241)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] aten codegen to filter backward ops / backends for inference-only mobile build"


Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure Android AAR build size:
```
scripts/build_pytorch_android.sh armeabi-v7a
```

This PR improved: 6.4M -> 5.8M
Baseline (before enabling c10 registers for static dispatch): 5.5M

Differential Revision: [D20266241](https://our.internmc.facebook.com/intern/diff/D20266241)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/gen.py:
```diff
@@ -202,6 +208,20 @@ def backend_to_devicetype(backend):
 }
 
 
+# Copied from tools.autograd.gen_python_functions.SKIP_PYTHON_BINDINGS
```
Can you add a reference here from `SKIP_PYTHON_BINDINGS` too?

ezyang commented:
This fails the sniff test for me. I could have written some code that calls one of these `_backward` functions directly (indeed, this is why https://github.com/pytorch/pytorch/blob/master/torch/nn/grad.py exists at all) and I don't see a moral justification why one *ought* to treat these specially; if binary size is the goal, you should be filtering out a lot more than backwards. I reviewed the internal discussion about the binary size regression and didn't see this question being asked.

ljk53 committed:
Update on "[pytorch] aten codegen to filter backward ops / backends for inference-only mobile build"


Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure Android AAR build size:
```
scripts/build_pytorch_android.sh armeabi-v7a
```

This PR improved: 6.4M -> 5.8M
Baseline (before enabling c10 registers for static dispatch): 5.5M

Differential Revision: [D20266241](https://our.internmc.facebook.com/intern/diff/D20266241)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] aten codegen to filter backward ops / backends for inference-only mobile build"


Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure Android AAR build size:
```
scripts/build_pytorch_android.sh armeabi-v7a
```

This PR improved: 6.4M -> 5.8M
Baseline (before enabling c10 registers for static dispatch): 5.5M

Differential Revision: [D20266241](https://our.internmc.facebook.com/intern/diff/D20266241)

[ghstack-poisoned]

ljk53 commented:
> This fails the sniff test for me. I could have written some code that calls one of these `_backward` functions directly (indeed, this is why https://github.com/pytorch/pytorch/blob/master/torch/nn/grad.py exists at all) and I don't see a moral justification why one _ought_ to treat these specially; if binary size is the goal, you should be filtering out a lot more than backwards. I reviewed the internal discussion about the binary size regression and didn't see this question being asked.

I totally agree this looks weird in aten/gen.py. The logic was originally introduced for PyTorch mobile 1.3 release in #26851 - at that time we didn't have custom build support yet so our goal was to shrink mobile binary release size as much as possible - mainly by assuming it's an inference-only library, but it's never a clean cut.

I think we have the following options:
1) What this PR does - having filtering logic in both aten/gen.py & tools/jit/gen_jit_dispatch.py for default mobile build. Very ugly, just to maintain the status quo during the migration;

2) Do nothing and regress OSS mobile default binary size by ~1M (5.5M->6.5M for armv7, +~4.5M for all 4 architectures). Developers who care about binary size can use custom build flow now, which doesn't regress much.

3) Make OSS mobile default build to be a special form custom build, i.e. instead of having this weird filtering logic in codegen we could have a separate util script to read native_functions.yaml, filtering out backward ops there and feed the result to custom build script. Or we can even add a new field `mobile_default_op` to native_functions.yaml. This way, it becomes the work to 'hand pick' relevant ops for default mobile binary, and is decoupled from core codegen logic.

ljk53 committed:
Update on "[pytorch] aten codegen to filter backward ops for inference-only mobile build"


Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure mobile build size:
```
// Default mobile build:
scripts/build_pytorch_android.sh armeabi-v7a

// MobileNetV2 custom build:
SELECTED_OP_LIST=MobileNetV2.yaml scripts/build_pytorch_android.sh armeabi-v7a

- arm-v7 Android AAR (compressed) size:
```
+----------+-------------------+---------------+
|          | MobileNetV2 Build | Default Build |
+----------+-------------------+---------------+
| Original |         3,354,589 |     5,731,992 |
| #34275   |         3,404,978 |     6,640,526 |
| #34622   |         3,432,569 |     6,640,526 |
| #34641   |         3,431,660 |     6,534,135 |
| This PR  |         3,429,480 |     6,049,387 |
+----------+-------------------+---------------+
```

Differential Revision: [D20266241](https://our.internmc.facebook.com/intern/diff/D20266241)

[ghstack-poisoned]

ezyang commented:
Both (2) and (3) seem reasonable. I think mobile users who are serious about space will need to use the custom build form anyway. (2) requires no extra work, right?

ljk53 committed:
Update on "[pytorch] aten codegen to filter backward ops for inference-only mobile build"


Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure mobile build size:
```
// Default mobile build:
scripts/build_pytorch_android.sh armeabi-v7a

// MobileNetV2 custom build:
SELECTED_OP_LIST=MobileNetV2.yaml scripts/build_pytorch_android.sh armeabi-v7a
```

- arm-v7 Android AAR (compressed) size:
```
+----------+-------------------+---------------+
|          | MobileNetV2 Build | Default Build |
+----------+-------------------+---------------+
| Original |         3,354,589 |     5,731,992 |
| #34275   |         3,404,978 |     6,640,526 |
| #34622   |         3,432,569 |     6,640,526 |
| #34641   |         3,431,660 |     6,534,135 |
| This PR  |         3,429,480 |     6,049,387 |
+----------+-------------------+---------------+
```

Differential Revision: [D20266241](https://our.internmc.facebook.com/intern/diff/D20266241)

[ghstack-poisoned]

ljk53 committed:
Update on "[pytorch] aten codegen to filter backward ops for inference-only mobile build"


Summary:
This is to address the mobile build size regression caused by moving c10
registers out of `#ifndef USE_STATIC_DISPATCH` in ATen/gen.py.

To measure mobile build size:
```
// Default mobile build:
scripts/build_pytorch_android.sh armeabi-v7a

// MobileNetV2 custom build:
SELECTED_OP_LIST=MobileNetV2.yaml scripts/build_pytorch_android.sh armeabi-v7a
```

- arm-v7 Android AAR (compressed) size:
```
+----------+-------------------+---------------+
|          | MobileNetV2 Build | Default Build |
+----------+-------------------+---------------+
| Original |         3,354,589 |     5,731,992 |
| #34275   |         3,404,978 |     6,640,526 |
| #34622   |         3,432,569 |     6,640,526 |
| #34641   |         3,431,660 |     6,534,135 |
| This PR  |         3,429,480 |     6,049,387 |
+----------+-------------------+---------------+
```

Differential Revision: [D20266241](https://our.internmc.facebook.com/intern/diff/D20266241)

[ghstack-poisoned]
